x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -0,0 +1,142 @@
+            if (version.before(Version.V_7_0_0_alpha1)) { // TODO(jaymode) change to V6_5_0 on backport!
+                LOGGER.warn("nodes prior to the minimum supported version for api keys {} exist in the cluster; these nodes will not be " +
+                    "able to use api keys", Version.V_7_0_0_alpha1);
+            }
+
+            final char[] keyHash = Hasher.PBKDF2.hash(apiKey);
+            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                builder.startObject()
+                    .field("doc_type", "api_key")
+                    .field("creation_time", created.toEpochMilli())
+                    .field("expiration_time", expiration == null ? null : expiration.toEpochMilli());
+
+                byte[] utf8Bytes = null;
+                try {
+                    utf8Bytes = CharArrays.toUtf8Bytes(keyHash);
+                    builder.field("api_key_hash").utf8Value(utf8Bytes, 0, utf8Bytes.length);
+                } finally {
+                    if (utf8Bytes != null) {
+                        Arrays.fill(utf8Bytes, (byte) 0);
+                    }
+                }
+
+                builder.array("role_descriptors", request.getRoleDescriptors())
+                    .field("name", request.getName())
+                    .field("version", version.id)
+                    .startObject("creator")
+                    .field("principal", authentication.getUser().principal())
+                    .field("metadata", authentication.getUser().metadata())
+                    .field("realm", authentication.getLookedUpBy() == null ?
+                        authentication.getAuthenticatedBy().getName() : authentication.getLookedUpBy().getName())
+                    .endObject()
+                    .endObject();
+                final IndexRequest indexRequest =
+                    client.prepareIndex(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, getApiKeyDocumentId(apiKey))
+                        .setOpType(DocWriteRequest.OpType.CREATE)
+                        .setSource(builder)
+                        .setRefreshPolicy(request.getRefreshPolicy())
+                        .request();
+                securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () ->
+                    executeAsyncWithOrigin(client, SECURITY_ORIGIN, IndexAction.INSTANCE, indexRequest,
+                        ActionListener.wrap(indexResponse ->
+                                listener.onResponse(new CreateApiKeyResponse(request.getName(), indexResponse.getId(), apiKey, expiration)),
+                            listener::onFailure)));
+            } catch (IOException e) {
+                listener.onFailure(e);
+            } finally {
+                Arrays.fill(keyHash, (char) 0);
+            }
+        }
+    }
+
+    private static String getApiKeyDocumentId(SecureString key) {
+        return "api_key_" + key;
```

Comment:
This means the document id still contains the raw API key. Shouldn't this use some other random identifier?